plugins/pulseaudio/PulseAudioDevice.cpp

```diff
@@ -15,28 +15,39 @@
  ******************************************************************************/
 
 #include "PulseAudioDevice.h"
+
+#include "Exception.h"
+#include "IReader.h"
 #include "PulseAudioLibrary.h"
+
 #include "devices/DeviceManager.h"
 #include "devices/IDeviceFactory.h"
-#include "Exception.h"
-#include "IReader.h"
 
 AUD_NAMESPACE_BEGIN
 
-PulseAudioDevice::PulseAudioSynchronizer::PulseAudioSynchronizer(PulseAudioDevice *device) :
-	m_device(device)
+PulseAudioDevice::PulseAudioSynchronizer::PulseAudioSynchronizer(PulseAudioDevice* device) : m_device(device)
 {
 }
 
-double PulseAudioDevice::PulseAudioSynchronizer::getPosition(std::shared_ptr<IHandle> handle)
+void PulseAudioDevice::PulseAudioSynchronizer::update_time_start()
 {
-	pa_usec_t latency;
-	int negative;
-	AUD_pa_stream_get_latency(m_device->m_stream, &latency, &negative);
```

Comment:
When you seek during playback you should still consider the latency/buffer size since the actual playback of the seeked handle will only happen once the already filled buffers are played back.

Reply:
I'm a bit unsure what you mean with this.

If you seek during playback and queue up a pause command immediately after, the reported position would be incorrect.

Same thing if you query the position directly after you seek. Lets say you seek to 0 and then the added latency would then make it report a negative seek position. This is exactly what this patch fixes as previously we needed special hacks in Blender to make sure that the actual reported position matched to what we actually were seeking to.

As this only happened in the pulseaudio backend, I don't think adding in latency to the reported position is the correct solution. Perhaps I'm misunderstanding you though?

Reply:
The old code computed the position as follows:

Get the position of the handle and then we subtract the following two values as the handle might not have been played back that far:

- Parts of what has been played back by the handle might be in the ring buffer waiting to be forwarded to pulse, so it hasn't been played yet.
- Pulse also has a latency which we can think of a buffer that is waiting to be played.

Of course, you're right that if we just seeked or if playback just started for a handle this will result in a time before the seek or a negative number if playback just started.

It is correct though in the sense that it tells you when it will start playing or start playing the position you seeked to. In that sense I don't consider this a bug, it's rather the most precise seeking in comparison to other backends which don't provide this latency information. It's the other backends that are less precise.

I understand that the timeline jumping back a bit in Blender is bad. This might happen of course, depending on the buffer size, the latency and also how the reported audio time is rounded to the blender playhead. Maybe a solution in Blender would be to seek to the middle of a frame, e.g., if rendering runs at 100 FPS and we seek to the second frame, we should seek to "frame" 2.5, i.e., t = 25 ms instead of t = 20 ms.